### ISSUE TITLE
rubocop: Array#sort! Too deep inlining, probably caused by recursive inlining.

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1727]
To reproduce:

```
git clone https://github.com/rubocop-hq/rubocop.git
cd rubocop
bundle
bundle exec rake spec
```

Error:

```
[truffle] opt fail         Array#sort! resource:/truffleruby/core/array.rb:1672        |Reason org.graalvm.compiler.core.common.PermanentBailoutException: Too deep inlining, probably caused by recursive inlining.
== Inlined methods ordered by inlining frequency:
java.util.DualPivotQuicksort.sort(int[], int, int, boolean) [977]
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(VirtualFrame) [2]
org.truffleruby.language.RubyRootNode.execute(VirtualFrame) [2]
org.truffleruby.language.control.SequenceNode.execute(VirtualFrame) [2]
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(VirtualFrame) [2]
org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(Object[]) [1]
org.truffleruby.language.dispatch.RubyCallNode.execute(VirtualFrame) [1]
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callRoot(Object[]) [1]
org.truffleruby.core.array.ArrayNodes$SortNode.sortPrimitiveArrayNoBlock(DynamicObject, NotProvided, ArrayStrategy, ArrayStrategy, ArrayOperationNodes$ArrayNewStoreNode, ArrayOperationNodes$ArrayCopyToNode, ArrayOperationNodes$ArraySortNode) [1]
org.truffleruby.language.dispatch.DispatchHeadNode.dispatch(VirtualFrame, Object, Object, DynamicObject, Object[]) [1]
org.truffleruby.builtins.InvokePrimitiveNode.execute(VirtualFrame) [1]
java.util.Arrays.sort(int[], int, int) [1]
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callInlined(Node, Object[]) [1]
org.truffleruby.language.dispatch.RubyCallNode.executeWithArgumentsEvaluated(VirtualFrame, Object, DynamicObject, Object[]) [1]
org.truffleruby.core.array.ArrayNodesFactory$StealArrayStorageNodeFactory$StealArrayStorageNodeGen.execute(VirtualFrame) [1]
java.util.DualPivotQuicksort.sort(int[], int, int, int[], int, int) [1]
org.truffleruby.core.array.IntegerArrayNodesFactory$ArraySortNodeGen.execute(Object, int) [1]
org.truffleruby.language.LazyRubyNode.execute(VirtualFrame) [1]
org.truffleruby.core.array.ArrayNodesFactory$SortNodeFactory$SortNodeGen.execute(VirtualFrame) [1]
org.truffleruby.language.methods.CatchForMethodNode.execute(VirtualFrame) [1]
org.truffleruby.language.dispatch.CachedDispatchNode.call(DirectCallNode, VirtualFrame, InternalMethod, Object, DynamicObject, Object[]) [1]
org.truffleruby.language.dispatch.CachedBoxedDispatchNode.executeDispatch(VirtualFrame, Object, Object, DynamicObject, Object[]) [1]
== Complete stack trace of inlined methods:
java.util.DualPivotQuicksort.sort(DualPivotQuicksort.java:215)
java.util.DualPivotQuicksort.sort(DualPivotQuicksort.java:539)
java.util.DualPivotQuicksort.sort(DualPivotQuicksort.java:539)
java.util.DualPivotQuicksort.sort(DualPivotQuicksort.java:539)
...
java.util.DualPivotQuicksort.sort(DualPivotQuicksort.java:111)
java.util.Arrays.sort(Arrays.java:169)
org.truffleruby.core.array.IntegerArrayNodesFactory$ArraySortNodeGen.execute(IntegerArrayNodesFactory.java:434)
org.truffleruby.core.array.ArrayNodes$SortNode.sortPrimitiveArrayNoBlock(ArrayNodes.java:2124)
org.truffleruby.core.array.ArrayNodesFactory$SortNodeFactory$SortNodeGen.execute(ArrayNodesFactory.java:8255)
org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:51)
org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:328)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callInlined(OptimizedCallTarget.java:268)
org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:63)
org.truffleruby.language.dispatch.CachedDispatchNode.call(CachedDispatchNode.java:190)
org.truffleruby.language.dispatch.CachedBoxedDispatchNode.executeDispatch(CachedBoxedDispatchNode.java:94)
org.truffleruby.language.dispatch.DispatchHeadNode.dispatch(DispatchHeadNode.java:43)
org.truffleruby.language.dispatch.RubyCallNode.executeWithArgumentsEvaluated(RubyCallNode.java:108)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:99)
org.truffleruby.core.array.ArrayNodesFactory$StealArrayStorageNodeFactory$StealArrayStorageNodeGen.execute(ArrayNodesFactory.java:8642)
org.truffleruby.builtins.InvokePrimitiveNode.execute(InvokePrimitiveNode.java:33)
org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
org.truffleruby.language.methods.CatchForMethodNode.execute(CatchForMethodNode.java:38)
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:51)
org.truffleruby.language.LazyRubyNode.execute(LazyRubyNode.java:42)
org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:328)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callInlined(OptimizedCallTarget.java:268)
org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:63)
org.truffleruby.language.dispatch.CachedDispatchNode.call(CachedDispatchNode.java:190)
org.truffleruby.language.dispatch.CachedBoxedDispatchNode.executeDispatch(CachedBoxedDispatchNode.java:94)
org.truffleruby.language.dispatch.DispatchHeadNode.dispatch(DispatchHeadNode.java:43)
org.truffleruby.language.dispatch.RubyCallNode.executeWithArgumentsEvaluated(RubyCallNode.java:108)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:99)
org.truffleruby.language.RubyNode.doExecuteVoid(RubyNode.java:54)
org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:31)
org.truffleruby.language.methods.CatchForMethodNode.execute(CatchForMethodNode.java:38)
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:51)
org.truffleruby.language.LazyRubyNode.execute(LazyRubyNode.java:42)
org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:328)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callInlined(OptimizedCallTarget.java:268)
org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:63)
org.truffleruby.language.dispatch.CachedDispatchNode.call(CachedDispatchNode.java:190)
org.truffleruby.language.dispatch.CachedBoxedDispatchNode.executeDispatch(CachedBoxedDispatchNode.java:94)
org.truffleruby.language.dispatch.DispatchHeadNode.dispatch(DispatchHeadNode.java:43)
org.truffleruby.language.dispatch.RubyCallNode.executeWithArgumentsEvaluated(RubyCallNode.java:108)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:99)
org.truffleruby.language.control.FrameOnStackNode.execute(FrameOnStackNode.java:35)
org.truffleruby.language.methods.CatchBreakNode.execute(CatchBreakNode.java:38)
org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
org.truffleruby.language.methods.CatchForMethodNode.execute(CatchForMethodNode.java:38)
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:51)
org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:328)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callInlined(OptimizedCallTarget.java:268)
org.graalvm.compiler.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:63)
org.truffleruby.language.dispatch.CachedDispatchNode.call(CachedDispatchNode.java:190)
org.truffleruby.language.dispatch.CachedBoxedDispatchNode.executeDispatch(CachedBoxedDispatchNode.java:94)
org.truffleruby.language.dispatch.DispatchHeadNode.dispatch(DispatchHeadNode.java:43)
org.truffleruby.language.dispatch.RubyCallNode.executeWithArgumentsEvaluated(RubyCallNode.java:108)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:99)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:79)
org.truffleruby.language.control.FrameOnStackNode.execute(FrameOnStackNode.java:35)
org.truffleruby.language.methods.CatchBreakNode.execute(CatchBreakNode.java:38)
org.truffleruby.language.dispatch.RubyCallNode.execute(RubyCallNode.java:79)
org.truffleruby.language.control.SequenceNode.execute(SequenceNode.java:34)
org.truffleruby.language.methods.CatchForMethodNode.execute(CatchForMethodNode.java:38)
org.truffleruby.language.methods.ExceptionTranslatingNode.execute(ExceptionTranslatingNode.java:51)
org.truffleruby.language.RubyRootNode.execute(RubyRootNode.java:54)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callProxy(OptimizedCallTarget.java:328)
org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callRoot(OptimizedCallTarget.java:318)
	at org.graalvm.compiler.replacements.PEGraphDecoder.tooDeepInlining(PEGraphDecoder.java:1103)
	at org.graalvm.compiler.replacements.PEGraphDecoder.doInline(PEGraphDecoder.java:869)
	at org.graalvm.compiler.replacements.PEGraphDecoder.tryInline(PEGraphDecoder.java:848)
	at org.graalvm.compiler.replacements.PEGraphDecoder.trySimplifyInvoke(PEGraphDecoder.java:722)
	at org.graalvm.compiler.replacements.PEGraphDecoder.handleInvoke(PEGraphDecoder.java:696)
	at org.graalvm.compiler.nodes.GraphDecoder.processNextNode(GraphDecoder.java:584)
	at org.graalvm.compiler.nodes.GraphDecoder.decode(GraphDecoder.java:415)
	at org.graalvm.compiler.replacements.PEGraphDecoder.decode(PEGraphDecoder.java:619)
	at org.graalvm.compiler.truffle.compiler.PartialEvaluator.doGraphPE(PartialEvaluator.java:482)
	at com.oracle.svm.truffle.api.SubstratePartialEvaluator.doGraphPE(SubstratePartialEvaluator.java:81)
	at org.graalvm.compiler.truffle.compiler.PartialEvaluator.fastPartialEvaluation(PartialEvaluator.java:520)
	at org.graalvm.compiler.truffle.compiler.PartialEvaluator.createGraph(PartialEvaluator.java:247)
	at org.graalvm.compiler.truffle.compiler.TruffleCompilerImpl.compileAST(TruffleCompilerImpl.java:441)
	at org.graalvm.compiler.truffle.compiler.TruffleCompilerImpl$TruffleCompilationWrapper.performCompilation(TruffleCompilerImpl.java:624)
	at org.graalvm.compiler.truffle.compiler.TruffleCompilerImpl$TruffleCompilationWrapper.performCompilation(TruffleCompilerImpl.java:571)
	at org.graalvm.compiler.core.CompilationWrapper.run(CompilationWrapper.java:178)
	at org.graalvm.compiler.truffle.compiler.TruffleCompilerImpl.doCompile(TruffleCompilerImpl.java:280)
	at org.graalvm.compiler.truffle.runtime.GraalTruffleRuntime.doCompile(GraalTruffleRuntime.java:687)
	at org.graalvm.compiler.truffle.runtime.GraalTruffleRuntime.doCompile(GraalTruffleRuntime.java:711)
	at org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$Request.run(BackgroundCompileQueue.java:83)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$TruffleCompilerThreadFactory$1.run(BackgroundCompileQueue.java:163)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:460)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)

.F..F...........FF..^C
```

**Comment by nirvdrum** [comment]
It looks like we're missing a `@TruffleBoundary` on a call into `Arrays.sort` and that call has unbounded recursion.

**Comment by deepj** [comment]
@eregon Hello, is there any new about  this issue?

**Comment by eregon** [comment]
This still seems an issue, I'll fix it.

**Comment by eregon** [comment]
Fixed by 5dc880e5421293f6c4d87db3fe9f578f2c709f07.
